sstables_loader.cc

```diff
@@ -203,11 +205,142 @@ class tablet_sstable_streamer : public sstable_streamer {
+    };
+    using sst_classification_info = std::vector<std::vector<minimal_sst_info>>;
+
+    future<>
+    stream_fully_contained_sstables(const dht::partition_range& pr, std::vector<sstables::shared_sstable> sstables, shared_ptr<stream_progress> progress) {
+        if (_stream_scope == stream_scope::node && _from_object_storage) {
+            llog.debug("Directly downloading {} fully contained SSTables to local node from object storage.", sstables.size());
+            return download_fully_contained_sstables(std::move(sstables), std::move(progress)).then([this](auto downloaded_ssts) -> future<> {
+                auto dwnld_ssts = std::move(downloaded_ssts);
+                auto dwnld_ssts_its = dwnld_ssts | std::views::transform([](auto& ssts) { return ssts.cbegin(); }) | std::ranges::to<std::vector>();
+                while (true) {
+                    std::vector<future<>> tasks;
+                    for (auto i = 0ul; i < dwnld_ssts_its.size(); ++i) {
+                        if (dwnld_ssts_its[i] != dwnld_ssts[i].end()) {
+                            tasks.emplace_back(smp::submit_to(
+                                i,
+                                [this, from = this_shard_id(), ks = _table.schema()->ks_name(), cf = _table.schema()->cf_name(), min_info = *dwnld_ssts_its[i]]
+                                -> future<> {
+                                    llog.debug("Adding downloaded SSTables to the table {} on shard {}, submitted from shard {}",
+                                               _table.schema()->cf_name(),
+                                               this_shard_id(),
+                                               from);
+                                    auto& db = _db.local();
+                                    auto& table = db.find_column_family(ks, cf);
+                                    auto& sst_manager = table.get_sstables_manager();
+                                    auto sst = sst_manager.make_sstable(table.schema(),
+                                                                                         table.get_storage_options(),
+                                                                                         min_info._generation,
+                                                                                         sstables::sstable_state::normal,
+                                                                                         min_info._version,
+                                                                                         min_info._format);
+                                    sst->set_sstable_level(0);
+                                    auto units = co_await sst_manager.dir_semaphore().get_units(1);
+                                    co_await sst->load(table.get_effective_replication_map()->get_sharder(*table.schema()));
+                                    co_await table.add_sstable_and_update_cache(sst);
+                                }));
```

Comment:
This is too deeply nested. Need to break up into named functions.

And, I think it should be part of storage (with a counterpart for local storage).

Reply:
We have two paths: break an sstable into fragments and stream it, or copy it. The copying part should be moved to storage, since it has one implementation for local sstables and another for remote sstables.